README.md
# [next] wp-scratch-box

<div style="background-color: #F8F9FA; color: #000; padding 4px; border: 0.10em solid red; border-radius: 2px">
⚠️ Warning: Development branch for V6.
</div>

<br>


![GitHub tag (latest by date)](https://img.shields.io/github/tag-date/apleasantview/wp-scratch-box.svg?label=release) [![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)

## ℹ️ Description
A quick and disposable Vagrant box for WordPress.   
It is suitable for a range of activities like presentations, workshops and development.

---

## 🔖 Table of Contents
- [Usage](#usage)
  * [Minimum Requirements](#minimum-requirements)
- [Provisioning](#provisioning)
  * [wp-scratch-box.sh](#wp-scratch-boxsh)
- [Configuration](#configuration)
  * [Vagrant.json](#vagrantjson)
    + [Default Vagrant configuration](#default-vagrant-configuration)
      - [Synced folder object:](#synced-folder-object-)
  * [WordPress Configuration](#wordpress-configuration)
    + [Default WordPress configuration](#default-wordpress-configuration)
    + [Optional parameters](#optional-parameters)
  * [LAMP default directory structure](#lamp-default-directory-structure)
  * [Resources (directory)](#resources--directory-)
  * [Database (directory)](#database--directory-)
  * [Scripts (directory)](#scripts--directory-)
  * [Vagrant plugins](#vagrant-plugins)
- [Configuration Warning](#configuration-warning)
- [Advanced Configuration](#advanced-configuration)
  * [Vagrant Multi-Machine](#vagrant-multi-machine)
    + [Muti-Machine configuration](#muti-machine-configuration)
  * [Multi-Machine Warning](#multi-machine-warning)
- [Contribute to `wp-scratch-box`](#contribute-to--wp-scratch-box-)
- [License](#license)

---

## ⌨️ Usage
```
git clone https://github.com/apleasantview/wp-scratch-box.git
cd wp-scratch-box
vagrant up
```
Visit `http://172.16.0.12` in your browser, you will be greeted by the five minute install - It's that easy! If you need a little bit more then read on below for details and configuration of your own `wp-scratch-box`.

### Minimum Requirements
- [Vagrant](https://www.vagrantup.com/) ( 1.7.4 > )
- [Virtualbox](https://www.virtualbox.org/) ( 5.0 > )

## 📜 Provisioning
### wp-scratch-box.sh
This is the provisioning file that will install the following packages and LAMP stack:
- **Packages:**
  - composer
  - curl
  - git-core
  - imagemagick
  - jq
  - ntp
  - software-properties-common
  - unzip
  - vim
  - zip
  - wp-cli w/ tab completions
- **LAMP**
  - Apache2
  - MariaDB 10.6
  - PHP-FPM 8.1
- **WordPress**
  - Latest stable version downloaded with WP-CLI.

##  🔧 Configuration
### Vagrant.json
Vagrant configuration can be modified in `Vagrant.json`, inside the `vagrant` object.

#### Default Vagrant configuration
| Config key | Vagrant setting | Default |
|-------------|---------|-------:|
| "name" | "vm.define", "vm.provider.name" | wp-scratch-box |
| "vagrant_box" | "vm.box" | ubuntu/bionic64 |
| "box_ip" | "vm.network" | 172.16.0.12 |
| "box_hostname" | "vm.hostname" | wp-scratch-box.test |
| "vb_cpus" | "vb.cpus" | 2 |
| "vb_memory" | "vb.memory" | 2048 |
| "vb_linked_clone" | "vb.linked_clone" | true |
| "synced_folder" | "vm.synced_folder"| <i>see below</i> |

##### Synced folder
| "synced_folder" | path |
| --------------- |:----:|
| "host_path" | "src/" |
| "guest_path" | "/var/www/public" |

Vagrant will create the `host_path` folder if it doesn't exist. The root directory will be synced to `/vagrant` per Vagrants' defaults, giving you access to any other folders and files needed. Note that there is no '/' at the end of the `guest_path`.

### WordPress Configuration
The latest stable version of WordPress is downloaded and configured through WP-CLI.

In `Vagrant.json` modify the values of the `wordpress` object:
```json
{
  "Project": {
    "vagrant": {
      // vagrant config
    },
    "wordpress": {
      "mysql_database": "wp_dummy",
      "mysql_user": "wp",
      "mysql_password": "wp",
      "mysql_prefix": "wp_"
    }
  }
}
```
#### Default WordPress configuration
These values will be read by the [jq utility](https://stedolan.github.io/jq/) at provisioning and passed on to WP-CLI and the `mysql` command to set up your database.

| Parameters | WP-CLI | Default | Set in Vagrant.json |
|:-----------|:------:|:-------:|---:|
| root_directory | *not used* | public | optional |
| core_directory | --path | **.** | optional |
| mysql_database | --dbname | wp_dummy | required |
| mysql_user | --dbuser | wp | required |
| mysql_password | --dbpass | wp | required |
| mysql_prefix | --dbprefix | wp_ | required |

#### Optional parameters
The `root_directory` and `core_directory` parameters are completely optional. These are mostly used in an advanced setup where WP Core is installed in its own directory. See this [Codex page](https://wordpress.org/support/article/giving-wordpress-its-own-directory/) for more info. 

### LAMP default directory structure
See above to modify where/how your site is served.

```
/var/www/  
+-- public/  
|	+-- .htaccess
|	+-- index.php
|	+-- wp-config.php
|	+-- wp-admin/
|	+-- wp-content/
|	+-- wp-include/
```

### Resources (directory)
Contains configuration files used during provisioning for Apache and PHP:  
`.htaccess` - `wp-scratch-box.conf` - `custom-php.ini`.
  
Please review these files and make adjustments accordingly if you change any default configuration values.

~~The `resources` folder also has a Mailcatcher installation script.~~ *Mailcatcher is deprecated. The script is still present but unlikely to work. It will be removed in the future for sure.*

### Database (directory)
A place to store your database(s). See README file in directory.

### Scripts (directory)
A place to story your custom scripts. See README file in directory.

### Vagrant plugins
If you have `vagrant-cachier` installed, the config in the Vagrantfile is set to cache by machine.
If you have `vagrant-vbguest` installed, guest additions updates is set to `false`. Manually update guest additions if really needed.

## ⚠️ Configuration Warning
If you should change the `root_directory` and `core_directory` parameters in `Vagrant.json`:
- Make sure `guest_path` is set accordingly in the `vagrant` configuration part of `Vagrant.json`.
- You will need to change paths in the apache configuration `wp-scratch-box.conf`, file is in the `resources` folder.
- Do not put a '/' at the front or end of these values.

## ⚒️ Advanced Configuration
### Vagrant Multi-Machine
By leveraging Vagrant's Multi-Machine feature, you can hook up an additional virtual machine or re-packaged boxes from VVV, Primary Vagrant, etc. alongside the main `Project` VM.

#### Muti-Machine configuration
In `Vagrant.json` append the following Parent Object:
```json
{
  "Project": {
    ...
  },
  "Custom": {
    "name": "test",
    "vagrant_box": "ubuntu/trusty64",
    "box_ip": "172.16.0.13",
    "box_hostname": ""
  }
}
```
Follow with command `vagrant up`

### ⚠️ Multi-Machine Warning 
- All keys are required and Parent Object must be set to `"Custom"` in `Vagrant.json`.
- The default `Project`machine will **not** start and be provisioned. You can force it to start with the command `vagrant up NAME`.
- If you run both `Project` and `Custom`, setting their IP's in the same range will create a network and facilitate communication between the two machines.
- Synced folders for the `Custom` machine needs to be manually set in the `Vagrantfile` if required.

***Please refer to the [Vagrant docs](https://docs.vagrantup.com/v2/multi-machine/index.html) for more info on Multi-Machine setups.***

## 🎉 Contribute to `wp-scratch-box`
- Open to any sort of contributions.
- Suggestions and improvements can be discussed in the issue tracker/through PR.

## ⚖️ License
Released under the MIT license.  
*2015 - 2019 a pleasant view | Cristovao Verstraeten*  
<br>

***Have a pleasant view!***


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [commit-and-tag-version](https://github.com/absolute-version/commit-and-tag-version) for commit guidelines.

## [5.4.0](https://github.com/apleasantview/wp-scratch-box/compare/v5.3.0...v5.4.0) (2023-02-24)


### Features

* LAMPS stack configuration update ([f7c08dc](https://github.com/apleasantview/wp-scratch-box/commit/f7c08dc739d7a9642a75f2227bad6806afbac27f))

## [5.3.0](https://github.com/apleasantview/wp-scratch-box/compare/v5.2.2...v5.3.0) (2022-11-17)


### Features

* **various:** Version bumps and memory increase ([c930b13](https://github.com/apleasantview/wp-scratch-box/commit/c930b13191c4a5a85b911af189fa12426bedeae0))


### Documentation

* README updated ([a1477e8](https://github.com/apleasantview/wp-scratch-box/commit/a1477e86effc05fb2650a6983776b62776c74862))


### Miscellaneous Chores

* Switch release utility ([ef0c71c](https://github.com/apleasantview/wp-scratch-box/commit/ef0c71c19fefcf12b1cb5dac8e3c01af466889cb))

### [5.2.2](https://github.com/apleasantview/wp-scratch-box/compare/v5.2.1...v5.2.2) (2021-04-06)


### Bug Fixes

* formatting error in versionrc ([9e0a490](https://github.com/apleasantview/wp-scratch-box/commit/9e0a490c5b660fb25de38d7faf944ecf6636e6a1))

### [5.2.1](https://github.com/apleasantview/wp-scratch-box/compare/v5.2.0...v5.2.1) (2021-04-06)


### Features

* Add note about host file config ([8699e68](https://github.com/apleasantview/wp-scratch-box/commit/8699e68ff821471772110bbcff11ceb1bb660dd5))
* **PHP:** Beef up PHP config values ([09b2e16](https://github.com/apleasantview/wp-scratch-box/commit/09b2e16cd1457cc66833a83641c9a3d68c4c52b0))


### Bug Fixes

* **Provisioning:** Imagick lib needed PHP version ([5b472fa](https://github.com/apleasantview/wp-scratch-box/commit/5b472fa912d01b6158169014382f4f6c2757bfdf))


### Miscellaneous Chores

* Update standard-version and rc formatting ([e7a0843](https://github.com/apleasantview/wp-scratch-box/commit/e7a0843b5fb613b9cadf794650c1ccd3836ac8e7))

## [5.2.0](https://github.com/apleasantview/wp-scratch-box/compare/v5.1.3...v5.2.0) (2021-01-14)


### Features

* **PHP:** Update PHP to 7.4 ([40a91a5](https://github.com/apleasantview/wp-scratch-box/commit/40a91a50b05beca17d34e1e4064e357e0fb1e38a))

### [5.1.3](https://github.com/apleasantview/wp-scratch-box/compare/v5.1.2...v5.1.3) (2021-01-14)

### [5.1.2](https://github.com/apleasantview/wp-scratch-box/compare/v5.1.1...v5.1.2) (2020-04-20)


### Miscellaneous Chores

* Update npm package ([27e85ae](https://github.com/apleasantview/wp-scratch-box/commit/27e85ae83daed3d64834d988fd203e3efea09ca9))

### [5.1.1](https://github.com/apleasantview/wp-scratch-box/compare/v5.1.0...v5.1.1) (2020-04-20)


### Bug Fixes

* **Provisioning:** Set restart services without asking to true ([25caac7](https://github.com/apleasantview/wp-scratch-box/commit/25caac7eec981eb013fbfc4671302e9084a9f79b))

## [5.1.0](https://github.com/apleasantview/wp-scratch-box/compare/v5.0.0...v5.1.0) (2019-11-07)


### Features

* Introduce db and scripts directory ([9b56257](https://github.com/apleasantview/wp-scratch-box/commit/9b56257802f4032215043817baa01aef923a5faf))


### Documentation

* Review and document new features ([1c9ccb7](https://github.com/apleasantview/wp-scratch-box/commit/1c9ccb7050bd02aeff6f32874b5c01d8199df99c))


### Code Refactoring

* **Provisioning:** Declare and assign separately parser variable in wp function ([da7527d](https://github.com/apleasantview/wp-scratch-box/commit/da7527d73474733849ccd08297aee1afa27b63f4))


### Miscellaneous Chores

* Ugrade standard-version ([0963eac](https://github.com/apleasantview/wp-scratch-box/commit/0963eac988f80c1f015783fbf8d2ef10ee89d979))
* Update project ignores ([698071a](https://github.com/apleasantview/wp-scratch-box/commit/698071a51d2fbf645c1f22f34dc3b9d3b81b312c))

## 5.0.0 (2019-07-20)


### Build System

* Introduce standard-version for versioning and releases ([8dcba3f](https://github.com/apleasantview/wp-scratch-box/commit/8dcba3f))


### Documentation

* Rewrite parts of the documentation following earlier changes ([98006a0](https://github.com/apleasantview/wp-scratch-box/commit/98006a0))


### Features

* **Vagrant:** Make configuration more verbose ([7b71154](https://github.com/apleasantview/wp-scratch-box/commit/7b71154))
* **WordPress:** Make configuration more verbose ([9bafdb9](https://github.com/apleasantview/wp-scratch-box/commit/9bafdb9))
* Log a more friendlier message when done provisioning ([3e34ff6](https://github.com/apleasantview/wp-scratch-box/commit/3e34ff6))


### BREAKING CHANGES

* **WordPress:** Make configuration more verbose
* **Vagrant:** Make configuration more verbose



## 4.1.0 (2019-05-25)



## 4.0.0 (2019-05-20)



## 3.1.0 (2016-08-24)



## 3.0.0 (2016-07-22)



### 2.10.8 (2016-07-10)



### 2.10.7 (2016-07-09)



### 2.10.4 (2016-06-29)



### 2.10.3 (2016-05-28)



## 2.10.0 (2016-05-22)



### 2.9.1 (2016-05-16)



## 2.8.0 (2016-04-18)



## 2.7.0 (2016-03-21)


resources/mailcatcher.sh
#!/bin/bash -
#title          :mailcatcher.sh
#description    :Installs MailCatcher on wp-scratch-box VM
#author         :Cristovao Verstraeten
#date           :20151202
#version        :20160629
#usage          :bash mailcatcher.sh
#notes          :
#bash_version   :4.3.39(3)-release
#============================================================================
set -euo pipefail
IFS=$'\n\t'

main(){
  # set -x
  required_packages
  rbenv_ruby_install
  mailcatcher_install
  # set +x
}

required_packages(){
  sudo apt-get purge -y \
    ruby
  sudo apt-get update && sudo apt-get install -y \
    autoconf \
    bison \
    build-essential \
    libffi-dev \
    libgdbm-dev \
    libgdbm3 \
    libncurses5-dev \
    libreadline6-dev \
    libsqlite3-dev \
    libssl-dev \
    libyaml-dev \
    openssl \
    zlib1g-dev
}

rbenv_ruby_install(){
  local ruby_version="2.3.0"
  
  git clone git://github.com/sstephenson/rbenv.git /home/vagrant/.rbenv
  git clone git://github.com/sstephenson/ruby-build.git /home/vagrant/.rbenv/plugins/ruby-build
  
  (
    cd /home/vagrant/.rbenv
    src/configure
    make -C src
  )
  
  {
    echo 'export PATH=$HOME/.rbenv/bin:$PATH'
    echo 'eval "$(rbenv init -)"'
  } >> /home/vagrant/.bash_profile
  source /home/vagrant/.bash_profile

  RUBY_CONFIGURE_OPTS=--disable-install-doc rbenv install "$ruby_version"
  rbenv global "$ruby_version"
  rbenv rehash
}

mailcatcher_install(){
  gem install mailcatcher --no-document
  rbenv rehash
  mailcatcher_php
  mailcatcher_upstart
  
  mailcatcher --ip=0.0.0.0 -v
}

mailcatcher_php(){
  local mailcatcher_path
  mailcatcher_path=$(rbenv which catchmail)

  printf "sendmail_path = %s\n" "$mailcatcher_path" | sudo tee /etc/php/5.6/mods-available/mailcatcher.ini
  sudo phpenmod mailcatcher
  sudo service php5.6-fpm restart; sudo service apache2 restart
}

mailcatcher_upstart(){
  local mailcatcher_path
  local ruby_path
  mailcatcher_path=$(rbenv which mailcatcher)
  ruby_path=$(rbenv which ruby)

  cat << EOF | sudo tee /etc/init/mailcatcher.conf
description "Mailcatcher"

start on runlevel [2345]
stop on runlevel [!2345]

respawn

exec $ruby_path $mailcatcher_path --foreground --http-ip=0.0.0.0
EOF
}

main

resources/ssl/ssl_setup.sh
#!/bin/bash -
set -euo pipefail
IFS=$'\n\t'

main(){
  # set -x
  write_keys
  apache2_configuration
  apache2_restart_services
  # set +x
}

write_keys() {
  sudo openssl req -x509 -nodes -newkey rsa:4096 \
    -keyout "wp-scratch-box_root.key" \
    -out "wp-scratch-box_root.crt" \
    -days 365 \
    -subj "/C=UK/O=wp-scratch-box Improved/CN=wp-scratch-box.test"
  sudo printf "[SAN]\nsubjectAltName=DNS:wp-scratch-box.test,DNS:www.wp-scratch-box.test\n" > wp-scratch-box.test.san
  sudo openssl genrsa -out "wp-scratch-box.test.key"
  sudo openssl req -new \
    -key "wp-scratch-box.test.key" \
    -out "wp-scratch-box.test.csr" \
    -subj "/CN=wp-scratch-box.test" \
    -days 365
  sudo openssl x509 -req -extfile "wp-scratch-box.test.san" -extensions "SAN" \
    -CAcreateserial \
    -CAserial "wp-scratch-box_root.srl" \
    -CAkey "wp-scratch-box_root.key" \
    -CA "wp-scratch-box_root.crt" \
    -in "wp-scratch-box.test.csr" \
    -out "wp-scratch-box.test.crt"
}

apache2_configuration(){
  sudo mkdir /etc/apache2/ssl
  sudo mv wp-scratch-box* /etc/apache2/ssl/
  sudo cp /vagrant/resources/ssl/ssl-params.conf /etc/apache2/conf-available/
  sudo cp /vagrant/resources/ssl/default-ssl.conf.bak /etc/apache2/sites-available/default-ssl.conf
}

apache2_restart_services(){
  sudo a2enmod ssl headers
  sudo a2ensite default-ssl
  sudo a2enconf ssl-params
  sudo apache2ctl configtest
  sudo systemctl restart apache2
}

main

wp-scratch-box.sh
#!/bin/bash -
#title          :wp-scratch-box.sh
#description    :provisioning script for wp-scratch-box
#author         :Cristovao Verstraeten
#date           :20151120
#version        :see CHANGELOG.md
#usage          :vagrant up --provision, vagrant provision
#notes          :
#bash_version   :4.3.39(3)-release
#============================================================================
set -euo pipefail
IFS=$'\n\t'

PHPVERSION=${PHPVERSION:-"8.1"}

main() {
  # set -x
  additional_repos
  base_packages
  lamp_install
  composer_install
  wordpress

  if [ -d "/home/vagrant/public" ];
    then
      echo "Symlink exists."
    else
      ln -s /var/www/public/ public
  fi
  # set +x
}

additional_repos() {
  # MariaDB.
  sudo curl -o /etc/apt/trusted.gpg.d/mariadb_release_signing_key.asc 'https://mariadb.org/mariadb_release_signing_key.asc'
  sudo sh -c "echo 'deb https://ftp.nluug.nl/db/mariadb/repo/10.6/ubuntu bionic main' >>/etc/apt/sources.list"

  # PHP and Apache by Ondrej Sury.
  sudo add-apt-repository ppa:ondrej/php
  sudo add-apt-repository ppa:ondrej/apache2
}

base_packages() {
  echo '* libraries/restart-without-asking boolean true' | sudo debconf-set-selections
  sudo apt-get update && sudo apt-get install -y \
    apt-transport-https \
    curl \
    git-core \
    imagemagick \
    jq \
    ntp \
    software-properties-common \
    unzip \
    vim \
    zip

  cp /home/vagrant/.profile /home/vagrant/.bash_profile
  wpcli_install
}

wpcli_install() {
  mkdir -p /home/vagrant/wp-cli
  (
    cd /home/vagrant/wp-cli
    curl -O -s -S https://raw.githubusercontent.com/wp-cli/builds/gh-pages/phar/wp-cli.phar
    curl -O -s -S https://raw.githubusercontent.com/wp-cli/wp-cli/master/utils/wp-completion.bash
    chmod +x wp-cli.phar
    sudo mv wp-cli.phar /usr/local/bin/wp
  )
  printf "\n%s\n" "source /home/vagrant/wp-cli/wp-completion.bash" >> .bash_profile
}

lamp_install() {
  apache_install
  mariadb_install
  phpfpm_install

  apache_configurations
}

apache_install() {
  sudo apt-get install -y \
    apache2
}

apache_configurations() {
  sudo a2enmod expires headers proxy proxy_fcgi rewrite setenvif
  sudo a2enconf php${PHPVERSION}-fpm
  sudo service apache2 restart

  sudo cp /vagrant/resources/apache/wp-scratch-box.conf /etc/apache2/sites-available/000-default.conf
  sudo service apache2 reload

  sudo usermod -a -G www-data vagrant
}

mariadb_install() {
  local root_password="root"

  echo "maria-db-10.6 mysql-server/root_password password $root_password" | sudo debconf-set-selections
  echo "maria-db-10.6 mysql-server/root_password_again password $root_password" | sudo debconf-set-selections
  sudo apt-get install -y mariadb-server

  # Revert to the old mysql_native_password authentication method. Needed from 10.4 on.
  local revert_auth=$( cat <<-SQL
ALTER USER 'root'@'localhost' IDENTIFIED WITH mysql_native_password USING PASSWORD('root');
FLUSH PRIVILEGES;
SQL
  )
  sudo mysql -u root -proot -e "${revert_auth}"

  # Run MySQL without passwords for convenience.
  (
    cat << 'EOF' | tee /home/vagrant/.my.cnf
[client]
user     = root
password = root

[mysqladmin]
user     = root
password = root
EOF
  )
}

phpfpm_install() {
  sudo apt-get install -y php${PHPVERSION}-fpm \
    php${PHPVERSION}-cli php${PHPVERSION}-common php${PHPVERSION}-bcmath php${PHPVERSION}-curl \
    php${PHPVERSION}-gd php${PHPVERSION}-imagick php${PHPVERSION}-imap php${PHPVERSION}-intl \
    php${PHPVERSION}-mbstring php${PHPVERSION}-mysql php${PHPVERSION}-pcov php${PHPVERSION}-soap \
    php${PHPVERSION}-ssh2 php${PHPVERSION}-xdebug php${PHPVERSION}-xml php${PHPVERSION}-xmlrpc php${PHPVERSION}-zip php-pear \
    php${PHPVERSION}-memcache php${PHPVERSION}-memcached

  sudo cp /vagrant/resources/php/php-fpm.conf /etc/php/${PHPVERSION}/fpm/php-fpm.conf
  sudo cp /vagrant/resources/php/pool-www.conf /etc/php/${PHPVERSION}/fpm/pool.d/www.conf
  sudo cp /vagrant/resources/php/custom-php.ini /etc/php/${PHPVERSION}/fpm/conf.d/php-custom.ini
  sudo cp /vagrant/resources/php/opcache.ini /etc/php/${PHPVERSION}/fpm/conf.d/opcache.ini
  # sudo phpenmod custom-php
  sudo phpdismod xdebug
  sudo phpdismod pcov

  # Explicitly restart PHP.
  sudo service php${PHPVERSION}-fpm restart &> /dev/null
}

composer_install() {
  curl -sS https://getcomposer.org/installer | sudo php -- --install-dir=/usr/local/bin --filename=composer
  composer --version

  # Fix 'permission denied' with vagrant-cachier symlink.
  if [ -d "/home/vagrant/.composer" ];
    then
      composer --version
      echo "Composer directory exists."
    else
      mkdir /home/vagrant/.composer
      sudo chown -R vagrant:vagrant /home/vagrant/.composer
  fi
}

wpcli_error_handler() {
  printf "WordPress files seem to already be present here. Moving on...\n"
  exit 0
}

wordpress() {
  local parser
  parser=$(jq -r 'if .Project.wordpress then .Project.wordpress|to_entries|map("\(.key)=\(.value|tostring)")| .[] else empty end' /vagrant/Vagrant.json)

  declare -A wp_custom
  while IFS="=" read -r key value; do
    wp_custom[$key]="$value";
  done < <(echo "$parser")

  local root_directory=${wp_custom[root_directory]:-public}
  local core_directory=${wp_custom[core_directory]:-.}
  local mysql_database=${wp_custom[mysql_database]:-wp_dummy}
  local mysql_user=${wp_custom[mysql_user]:-wp}
  local mysql_password=${wp_custom[mysql_password]:-wp}
  local mysql_prefix=${wp_custom[mysql_prefix]:-wp_}

  mysql -u root -e "CREATE DATABASE IF NOT EXISTS $mysql_database;"
  mysql -u root -e "GRANT ALL PRIVILEGES ON $mysql_database.* TO $mysql_user@localhost IDENTIFIED BY '$mysql_password';"

  mkdir -p "/var/www/$root_directory"
  (
    trap 'wpcli_error_handler' ERR

    cd "/var/www/$root_directory"
    wp cli version
    wp core download --path="$core_directory/" 2> /dev/null
    wp core config --path="$core_directory/" --dbname="$mysql_database" --dbuser="$mysql_user" --dbpass="$mysql_password" --dbprefix="$mysql_prefix"
    wp core version --path="$core_directory/" --extra
  )
}

main

printf "\n========\n"
printf "\n\tProvisioning script finished without errors.\n"
printf "\tVisit http://172.16.0.12 for the 5 minute install!\n"
printf "\tIf you would like to access the site from a custom URL, add it to your hosts file.\n"
printf "\t\tHave a pleasant view!\n"
echo " "
